### PR TITLE
Don't cache CanReuse value

### LINKED ISSUE #34768: Kestrel reset error when no reset received
RE: https://github.com/dotnet/runtime/issues/55261

The following logs show kestrel throwing a `The client reset the request stream.` error when no reset frame was received for that stream. These tests started failing between between 6.0 Preview 4 and 6.0 Preview 6, but we don't have server logs from Preview 4 to confirm the prior behavior.
```
2021-07-27T17:28:48.0481734+02:00  [VRB] Connection id ""0HMAH2CT4I8BM"" received HEADERS frame for stream ID 1831 with length 26 and flags END_HEADERS. (965cec00)

2021-07-27T17:28:48.0482203+02:00  [VRB] Connection id ""0HMAH2CT4I8BM"" received DATA frame for stream ID 1831 with length 4270 and flags NONE. (965cec00)

2021-07-27T17:28:48.0482744+02:00 0HMAH2CT4I8BM:00000727 [INF] Request starting HTTP/2 POST https://localhost:5001/ - 4270 (ca22a1cb)
2021-07-27T17:28:48.0482962+02:00 0HMAH2CT4I8BM:00000727 [VRB] All hosts are allowed. (2e40afda)

2021-07-27T17:28:48.0483186+02:00  [VRB] Connection id ""0HMAH2CT4I8BM"" received DATA frame for stream ID 1831 with length 0 and flags END_STREAM. (965cec00)
2021-07-27T17:28:48.0483201+02:00 0HMAH2CT4I8BM:00000727 [DBG] 1 candidate(s) found for the request path '"/"' (9406aaa8)
2021-07-27T17:28:48.0483411+02:00 0HMAH2CT4I8BM:00000727 [DBG] Request matched endpoint '"/ HTTP: POST"' (cbf60c4b)
2021-07-27T17:28:48.0483583+02:00 0HMAH2CT4I8BM:00000727 [INF] Executing endpoint '"/ HTTP: POST"' (500cc934)

2021-07-27T17:28:48.0485281+02:00 0HMAH2CT4I8BM:00000727 [INF] Executed endpoint '"/ HTTP: POST"' (99874f2b)

2021-07-27T17:28:48.0486058+02:00 0HMAH2CT4I8BM:00000727 [ERR] Connection id ""0HMAH2CT4I8BM"", Request id ""0HMAH2CT4I8BM:00000727"": An unhandled exception was thrown by the application. (560e7d32)
System.IO.IOException: The client reset the request stream.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestPipeReader.ReadAsync(CancellationToken cancellationToken) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000b2b+0x0
   at System.IO.Pipelines.PipeReader.CopyToAsyncCore[TStream](TStream destination, Func`4 writeAsync, CancellationToken cancellationToken) in System.IO.Pipelines.dll:token 0x60000b8+0x18
   at HttpStress.StressServer.<>c.<<MapRoutes>b__6_6>d.MoveNext() in c:\repos\runtime\src\libraries\System.Net.Http\tests\StressTests\HttpStress\StressServer.cs:line 262
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger) in Microsoft.AspNetCore.Routing.dll:token 0x60000ab+0x5e
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000aa1+0x1b8
2021-07-27T17:28:48.0486522+02:00 0HMAH2CT4I8BM:00000727 [VRB] Connection id ""0HMAH2CT4I8BM"" sending HEADERS frame for stream ID 1831 with length 7 and flags END_STREAM, END_HEADERS. (5439505f)
2021-07-27T17:28:48.0487189+02:00 0HMAH2CT4I8BM:00000727 [INF] Request finished HTTP/2 POST https://localhost:5001/ - 4270 - 500 0 - 0.4368ms (791a596a)
2021-07-27T17:28:48.0487483+02:00  [DBG] Connection id ""0HMAH2CT4I8BM"", Request id ""0HMAH2CT4I8BM:00000727"": started reading request body. (55b8f554)
2021-07-27T17:28:48.0487667+02:00  [DBG] Connection id ""0HMAH2CT4I8BM"", Request id ""0HMAH2CT4I8BM:00000727"": done reading request body. (6feb3b06)
```

Theory: Kestrel is caching an exception in the HttpRequestPipeReader from a previously reset stream instance, the body control state is never cleared after aborted.
https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L503

However, caching this shouldn't be possible if CanReuse correctly guards against caching an aborted stream.
https://github.com/dotnet/aspnetcore/blob/39912f8f73e06f5d9735745d24545aed8821698f/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs#L169

There may be race conditions between the RESET_STREAM frame processing and the stream cleanup logic for caching streams.

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #34768

## Regression?
- [ ] Yes
- [x] No

## Risk
- [ ] High
- [ ] Medium
- [x] Low

Worst case is that Http2 streams are pooled less often.

## Verification
- [X] Manual (required)
- [ ] Automated

## Packaging changes reviewed?
- [ ] Yes
- [ ] No
- [X] N/A

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs [hunk 1]
```diff
@@ -107,7 +106,9 @@ public bool ReceivedEmptyRequestBody
             }
         }
 
-        public bool CanReuse { get; private set; }
+        // We only want to reuse a stream that was not aborted and has completely finished writing.
+        // This ensures Http2OutputProducer.ProcessDataWrites is in the correct state to be reused.
+        public bool CanReuse => !_connectionAborted && HasResponseCompleted;
 
         protected override void OnReset()
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs:113]: Would resetting the value to `false`  in OnReset accomplish the same goal?

**Review by JamesNK — CHANGES_REQUESTED** [review]
Please add a unit test. This is subtle behavior and we want to ensure it isn't regressed in the future.

Where is `_connectionAborted` set to true? Will it always happen before the stream is pooled? I want to make sure there is no chance that a stream is pooled and then some code aborts it while it is in the pool.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs (hunk 1 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs:113]: The line on 44 was effectively doing that before. The issue was that `!_connectionAborted && HasResponseCompleted` was being evaluated too early on a background thread and a `RST_STREAM` frame could still be processed and abort the stream before pooling. Now `CanReuse` is evaluated once in the main frame-processing loop right before it's pooled or not.

**Review by halter73 — APPROVED** [review]

**Comment by sebastienros** [comment]
> Where is _connectionAborted set to true?

When the RST_STREAM is processed.

**Comment by halter73** [comment]
> Please add a unit test.

I would also love that if you find a good way to do it @JamesNK. The problem is the `RST_STREAM` frame needs to be processed between `Http2Stream.CompleteStream(bool)` executing on a background thread and `Http2Connection.UpdateCompletedStreams()` running again.

This would be easy enough to do by adding some sort of event/resettable-Task we could hook in the tests when `Http2Connection.OnStreamCompleted()` is called and only setting the `RST_STREAM` frame then, but the problem there is that `UpdateCompletedStreams()` runs immediately after `await Input.ReadAsync()` completes without processing any incoming frames. So we'd also need a way to force processing frames without calling `UpdateCompletedStreams()` to simulate Http2Connection frame-processing loop running concurrently with `Http2Stream.CompleteStream(bool)`. I'm not sure it's worth it.

**Comment by JamesNK** [comment]
Ok. If we can't add a unit test then we should comment. Something like:

> CanReuse be evaluated on the main frame-processing looping after the stream is removed from the connection's active streams collection. This is required because a RST_STREAM frame could arrive after the END_STREAM flag is received. Once the stream is no longer in the connection's active stream collection then it can no longer be reset and is safe to evaluate for pooling.

**Comment by davidfowl** [comment]
Nice catch!

**Comment by davidfowl** [comment]
Why is this is main and not rc1?

**Comment by sebastienros** [comment]
@davidfowl merge to main, then backport to rc1 (standard procedure I believe)

**Review by JamesNK — APPROVED** [review]

**Pull request merged**